src/Microsoft.Windows.CsWin32/Generator.cs

```diff
@@ -3137,7 +3159,7 @@ private ParameterSyntax CreateParameter(MethodSignature<TypeSyntax> methodSignat
                     //     internal Span<TheStruct> AsSpan() => MemoryMarshal.CreateSpan(ref _1, 4);
```

Comment:
Can `internal ref TheStruct this[int index] => ref AsSpan()[index];` be made an extension method too, so that the current C# compiler tracks the ref lifetime correctly?

In fact, why not always emit the same extension method impl for it whether or not AsSpan exists?

Reply:
No, proper indexers cannot be written as extension methods AFAIK. 
This indexer only exists where a `Span<T>` member can legitimately exist as well, so the ref lifetime should be fine, shouldn't it?

> In fact, why not always emit the same extension method impl for it whether or not AsSpan exists?

We do. Notice the tests which show the extension methods are always generated. On netcoreapp3.1+ the extension methods are not necessary, but I keep them so folks who multitarget can write using just the extension methods so it's just one codebase.

Reply:
My understanding was that the Span member is not legitimate and can't be, from the standpoint of tracking the lifetime of the ref. We fixed the GC hole issue, but we intentionally left the issue that the span can outlive the memory it points to. Same problem with the indexer.

Reply:
From https://github.com/microsoft/CsWin32/issues/117#issuecomment-785192927

Reply:
So not having an indexer and having the C#-9-safe extension method seems better to me because it doesn't come with the outliving problem that the indexer and the AsSpan member bring.

Reply:
If the indexer and `AsSpan` method have clear XML documentation that explains that the ref can live longer than the memory it points to, and be left pointing at arbitrary memory, then I don't mind them. `CreateUnsafeSpan` might be a better name too.

Reply:
This seems out of scope of the PR since since it only adds extension methods. But I will ask Jared about it and open a new PR if fixing the existing instance members is required. Thanks for calling it out.

Reply:
> We do. Notice the tests which show the extension methods are always generated. On netcoreapp3.1+ the extension methods are not necessary, but I keep them so folks who multitarget can write using just the extension methods so it's just one codebase.

This answers my main concern. I missed that.

If the indexer and `AsSpan` method have clear XML documentation that explains that the ref and the span can live longer than the memory they point to, and be left pointing at arbitrary memory if you don't keep the target alive, then I don't mind them. `CreateUnsafeSpan` might be a better name too.

What if you make the indexer get and set directly instead of returning a ref? That way you have the nice syntax for the 99% case, but you don't have the lifetime mismatch.

Reply:
In other words:

```cs
internal TheStruct this[int index]
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    get => this.ItemRef(index);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    set => this.ItemRef(index) = value;
}
```